### PR TITLE
[FEATURE] Benchmark Search Module

### PR DESCRIPTION
**Pull request by svnbgnk**

### CODE REVIEW

File: test/performance/search/search_benchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+#include <benchmark/benchmark.h>
+
+#include <seqan3/alphabet/aminoacid/aa20.hpp>
+#include <seqan3/alphabet/nucleotide/dna4.hpp>
+#include <seqan3/search/algorithm/all.hpp>
+//
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]: remove

File: test/performance/search/search_benchmark.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    text.resize(length);
+    for (unsigned i = 0; i < length; ++i)
+    {
+        alphabet_t r;
+        r.assign_rank(std::rand() % alphabet_size);
+        text[i] = r;
+    }
+}
+
+template<typename alphabet_t>
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
template <Alphabet alphabet_t>
```

File: test/performance/search/search_benchmark.cpp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+void mutate_insertion(std::vector<alphabet_t> & seq, size_t const  errors){
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
void mutate_insertion(std::vector<alphabet_t> & seq, size_t const errors){
```

File: test/performance/search/search_benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    size_t const alphabet_size = alphabet_t::value_size;
+    size_t rpos = rand() % (seq.size() - errors);
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
    size_t rpos = rand() % (std::ranges::size(seq) - errors);
```

File: test/performance/search/search_benchmark.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    size_t rrank = rand() % alphabet_size;
+    alphabet_t cbase;
+    cbase.assign_rank(rrank);
+    seq.insert(seq.begin() + rpos, cbase);
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
    seq.insert(seq.begin() + rpos, alphabet_t{}.assign_rank(rrank));
```

Should work. Then you can remove the two lines above

File: test/performance/search/search_benchmark.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+template<typename alphabet_t>
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
template <Alphabet alphabet_t>
```

File: test/performance/search/search_benchmark.cpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+void mutate_deletion(std::vector<alphabet_t> & seq, size_t const  errors){
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
void mutate_deletion(std::vector<alphabet_t> & seq, size_t const errors){
```

File: test/performance/search/search_benchmark.cpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    uint32_t rpos = rand() % (seq.size() - errors);
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
    size_t rpos = rand() % (std::ranges::size(seq) - errors);
```

File: test/performance/search/search_benchmark.cpp [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    seq.erase(seq.begin() + rpos);
+}
+
+template<typename alphabet_t>
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
template <Alphabet alphabet_t>
```

File: test/performance/search/search_benchmark.cpp [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+void mutate_substitution(std::vector<alphabet_t> & seq, size_t const  errors){
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
void mutate_substitution(std::vector<alphabet_t> & seq, size_t const errors){
```

File: test/performance/search/search_benchmark.cpp [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    size_t const alphabet_size = alphabet_t::value_size;
+    size_t rpos = rand() % (seq.size() - errors);
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
    size_t rpos = rand() % (std::ranges::size(seq) - errors);
```

File: test/performance/search/search_benchmark.cpp [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        alphabet_t r;
+        r.assign_rank(std::rand() % alphabet_size);
+        text[i] = r;
+    }
+}
+
+template<typename alphabet_t>
+void mutate_insertion(std::vector<alphabet_t> & seq, size_t const  errors){
+    size_t const alphabet_size = alphabet_t::value_size;
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
    constexpr size_t alphabet_size = alphabet_size_v<alphabet_t>; 
```

File: test/performance/search/search_benchmark.cpp [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+template<typename alphabet_t>
+void mutate_deletion(std::vector<alphabet_t> & seq, size_t const  errors){
+    uint32_t rpos = rand() % (seq.size() - errors);
+    seq.erase(seq.begin() + rpos);
+}
+
+template<typename alphabet_t>
+void mutate_substitution(std::vector<alphabet_t> & seq, size_t const  errors){
+    size_t const alphabet_size = alphabet_t::value_size;
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
    constexpr size_t alphabet_size = alphabet_size_v<alphabet_t>; 
```

File: test/performance/search/search_benchmark.cpp [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    size_t rpos = rand() % (seq.size() - errors);
+    size_t rrank = rand() % (alphabet_size - 1);
+    alphabet_t & cbase = seq[rpos];
+    size_t crank = to_rank(cbase);
+    if(rrank >=  crank)
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]: Don't you need to re-simulate a new base until `rrank != crank` ?

File: test/performance/search/search_benchmark.cpp [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        ++rrank;
+    cbase.assign_rank(rrank);
+}
+
+template<typename alphabet_t>
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
template <Alphabet alphabet_t>
```

File: test/performance/search/search_benchmark.cpp [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+void generate_reads(std::vector<std::vector<alphabet_t> > & reads,
+                    std::vector<alphabet_t> & ref,
+                    size_t const number_of_Reads,
+                    size_t const read_length,
+                    size_t const simulated_errors,
+                    float const prob_insertion,
+                    float const prob_deletion
+                   )
+{
+    for(size_t i = 0; i < number_of_Reads; ++i){
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
    for (size_t i = 0; i < number_of_Reads; ++i){
```

File: test/performance/search/search_benchmark.cpp [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+    size_t crank = to_rank(cbase);
+    if(rrank >=  crank)
+        ++rrank;
+    cbase.assign_rank(rrank);
+}
+
+template<typename alphabet_t>
+void generate_reads(std::vector<std::vector<alphabet_t> > & reads,
+                    std::vector<alphabet_t> & ref,
+                    size_t const number_of_Reads,
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
                    size_t const number_of_reads,
```

File: test/performance/search/search_benchmark.cpp [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+                    size_t const simulated_errors,
+                    float const prob_insertion,
+                    float const prob_deletion
+                   )
+{
+    for(size_t i = 0; i < number_of_Reads; ++i){
+        size_t rpos = rand() % (ref.size() - read_length - simulated_errors);
+        std::vector<alphabet_t> read_tmp{ref.begin() + rpos,
+            ref.begin() + rpos + read_length + simulated_errors};
+        for(size_t j = 0; j < simulated_errors; ++j)
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
        for (size_t j = 0; j < simulated_errors; ++j)
```

File: test/performance/search/search_benchmark.cpp [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+        {
+            //Substitution
+            float prob = (float) rand()/RAND_MAX;
+            if(prob_insertion + prob_deletion < prob)
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
            if (prob_insertion + prob_deletion < prob)
```

File: test/performance/search/search_benchmark.cpp [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+            {
+                mutate_substitution(read_tmp, simulated_errors);
+            }
+            //Insertion
+            else if(prob_insertion < prob)
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
            else if (prob_insertion < prob)
```

File: test/performance/search/search_benchmark.cpp [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+            {
+                mutate_insertion(read_tmp, simulated_errors);
+            }
+            //Deletion
+            else
+            {
+                mutate_deletion(read_tmp, simulated_errors);
+            }
+        }
+        std::vector<alphabet_t>  read{read_tmp.begin(), read_tmp.begin() + read_length};
```
Comment (eseiler) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
        std::vector<alphabet_t> read{read_tmp.begin(), read_tmp.begin() + read_length};
```

File: test/performance/search/search_benchmark.cpp (hunk 14 above)
Comment (svnbgnk) [reply test/performance/search/search_benchmark.cpp]: No, for dna4 I get the random numbers 0, 1, 2. (note 3 is missing)
If the current letter has the rank 1 I just shift the random numbers 1, 2 -> 2, 3.
This way I ensure there is always a new letter at this place.

Comment (eseiler) [reply test/performance/search/search_benchmark.cpp]: Ah right :)

File: test/performance/search/search_benchmark.cpp [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+
+//============================================================================
+//  undirectional; trivial_search, single, dna4, searched_error = 3
+//============================================================================
+
+void unidirectional_search(benchmark::State & state)
+{
+    uint8_t const simulated_errors = state.range(0);
+    uint8_t const searched_errors = state.range(1);
+    size_t reference_length{100'000};
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]: You can make all variables `const`.

Is there a reason why you use the assignment operator for `uint8_t`?

File: test/performance/search/search_benchmark.cpp [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+    int number_of_reads{100};
+    size_t read_length{100};
+    float prob_insertion{0.18};
+    float prob_deletion{0.18};
+
+    std::vector<seqan3::dna4> ref = generate_sequence_seqan3<seqan3::dna4>(reference_length);
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]: I would suggest to replace everything with `dna4_vector` and remove `seqan3::` namespace references to make it more readable.

File: test/performance/search/search_benchmark.cpp [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+
+template<Alphabet alphabet_t>
+void mutate_deletion(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0){
+    std::mt19937 gen(seed);
+    std::uniform_int_distribution<size_t> random_pos(0, std::ranges::size(seq) - overlap);
+    seq.erase(seq.begin() + random_pos(gen));
+}
+
+template<Alphabet alphabet_t>
+void mutate_substitution(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0){
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
void mutate_substitution(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0)
{
```

File: test/performance/search/search_benchmark.cpp [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+void mutate_insertion(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0){
+    std::mt19937 gen(seed);
+    std::uniform_int_distribution<uint8_t> dis_alpha(0, alphabet_size_v<alphabet_t> - 1);
+    std::uniform_int_distribution<size_t> random_pos(0, std::ranges::size(seq) - overlap);
+    alphabet_t cbase;
+    seq.insert(seq.begin() + random_pos(gen), alphabet_t{}.assign_rank(dis_alpha(gen)));
+}
+
+template<Alphabet alphabet_t>
+void mutate_deletion(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0){
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
void mutate_deletion(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0)
{
```

File: test/performance/search/search_benchmark.cpp [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+
+    size_t length = dis_length(gen);
+    for (size_t l = 0; l < length; ++l)
+        sequence.push_back(alphabet_t{}.assign_rank(dis_alpha(gen)));
+
+    return sequence;
+}
+
+template<Alphabet alphabet_t>
+void mutate_insertion(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0){
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
void mutate_insertion(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0)
{
```

File: test/performance/search/search_benchmark.cpp [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+    std::vector<alphabet_t> sequence;
+
+    size_t length = dis_length(gen);
+    for (size_t l = 0; l < length; ++l)
+        sequence.push_back(alphabet_t{}.assign_rank(dis_alpha(gen)));
+
+    return sequence;
+}
+
+template<Alphabet alphabet_t>
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
template <Alphabet alphabet_t>
```

File: test/performance/search/search_benchmark.cpp [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+void mutate_insertion(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0){
+    std::mt19937 gen(seed);
+    std::uniform_int_distribution<uint8_t> dis_alpha(0, alphabet_size_v<alphabet_t> - 1);
+    std::uniform_int_distribution<size_t> random_pos(0, std::ranges::size(seq) - overlap);
+    alphabet_t cbase;
+    seq.insert(seq.begin() + random_pos(gen), alphabet_t{}.assign_rank(dis_alpha(gen)));
+}
+
+template<Alphabet alphabet_t>
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
template <Alphabet alphabet_t>
```

File: test/performance/search/search_benchmark.cpp [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+void mutate_deletion(std::vector<alphabet_t> & seq, size_t const overlap, size_t const seed = 0){
+    std::mt19937 gen(seed);
+    std::uniform_int_distribution<size_t> random_pos(0, std::ranges::size(seq) - overlap);
+    seq.erase(seq.begin() + random_pos(gen));
+}
+
+template<Alphabet alphabet_t>
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
template <Alphabet alphabet_t>
```

File: test/performance/search/search_benchmark.cpp [hunk 30]
```diff
@@ -0,0 +1,142 @@
...
+    std::uniform_int_distribution<size_t> random_pos(0, std::ranges::size(seq) - overlap);
+    alphabet_t & cbase = seq[random_pos(gen)];
+    uint8_t crank = to_rank(cbase);
+    uint8_t rrank = dis_alpha_short(gen);
+    if (rrank >=  crank)
+        ++rrank;
+    cbase.assign_rank(rrank);
+}
+
+template<Alphabet alphabet_t>
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
template <Alphabet alphabet_t>
```

File: test/performance/search/search_benchmark.cpp [hunk 31]
```diff
@@ -0,0 +1,142 @@
...
+                    float const prob_insertion,
+                    float const prob_deletion,
+                    size_t const seed = 0)
+{
+    std::vector<std::vector<alphabet_t> > reads;
+    std::mt19937 gen(seed);
+    std::uniform_int_distribution<size_t> seeds (0, SIZE_MAX);
+    std::uniform_int_distribution<size_t> random_pos(0, std::ranges::size(ref) - read_length - simulated_errors);
+    std::uniform_real_distribution<double> probability(0.0, 1.0);
+    for (size_t i = 0; i < number_of_reads; ++i){
```
Comment (cpockrandt) [inline test/performance/search/search_benchmark.cpp]:
```suggestion
    for (size_t i = 0; i < number_of_reads; ++i)
    {
```

**Comment by h-2** [comment]
What's the status here?

**Comment by svnbgnk** [comment]
I am waiting for a globally available generate text function (for all benchmarks).

**Comment by marehr** [comment]
@svnbgnk Can you rebase?

**Comment by h-2** [comment]
@svnbgnk Can you rebase this against current master and clean-up the commit history?

**Comment by h-2** [comment]
Whats the status here?

**Comment by svnbgnk** [comment]
Hm something went wrong during the rebase? The easy solution for me is to just create a new PR.

**Comment by eseiler** [comment]
Superseded by #1100

**Pull request closed without merging**